src/firmware/app/control/trajectory_planner.c

```diff
@@ -1,275 +1,241 @@
+
+    // Generate the segment-based duration of each trajectory
+    app_trajectory_planner_generatePositionTrajectoryTimeProfile(
+        linear_segment_lengths, linear_speed, linear_time_profile, num_elements);
+    app_trajectory_planner_generatePositionTrajectoryTimeProfile(
+        angular_segment_lengths, angular_speed, angular_time_profile, num_elements);
+
+    // Calculate the time duration of the trajectory at each segment node
+    app_trajectory_planner_modifySpeedsToMatchDuration(
+        linear_speed, angular_speed, linear_time_profile, angular_time_profile,
+        linear_segment_lengths, angular_segment_lengths,
+        position_trajectory->time_profile, num_elements);
 
-    // Set the new number of time periods
-    constant_period_trajectory->path_parameters.num_segments = ++time_periods;
     return OK;
 }
 
-static void app_trajectory_planner_generateConstArclengthTrajectoryPositions(
-    PositionTrajectoryElement_t traj_elements[TRAJECTORY_PLANNER_MAX_NUM_ELEMENTS],
-    Polynomial2dOrder3_t path, const unsigned int num_elements,
-    ArcLengthParametrization_t arc_length_parameterization,
-    const float arc_segment_length)
+void app_trajectory_planner_generatePositionTrajectoryTimeProfile(
+    float* segment_lengths, float* speeds, float* trajectory_durations,
+    unsigned int num_elements)
 {
+    // Calculate the time required to move between the first and last nodes of a
+    // trajectory segment
     for (unsigned int i = 0; i < num_elements; i++)
     {
-        // Get the 't' value corresponding to the current arc length (to be used for
-        // further computing)
-        float t = shared_polynomial2d_getTValueAtArcLengthOrder3(
-            path, i * arc_segment_length, arc_length_parameterization);
+        // Delta-time over the length of the segment
+        float delta_time = 0;
 
-        // Get the X and Y position at the 't' value defined by the arc length
-        traj_elements[i].position = shared_polynomial2d_getValueOrder3(path, t);
+        // Check that we are not dividing by zero
+        if (speeds[i + 1] == 0 && speeds[i] == 0)
+        {
+            delta_time = 0;
+        }
+        else
+        {
+            delta_time = (2 * segment_lengths[i]) / (speeds[i + 1] + speeds[i]);
+        }
+
+        // Catch-all case that there is no change in position for both the linear and
+        // angular paths
+        trajectory_durations[i] = delta_time;
     }
 }
 
 void app_trajectory_planner_getMaxAllowableSpeedProfile(
-    float max_allowable_speed_profile[TRAJECTORY_PLANNER_MAX_NUM_ELEMENTS],
-    Polynomial2dOrder3_t path, unsigned int num_elements,
-    ArcLengthParametrization_t arc_length_parameterization,
-    const float arc_segment_length, const float max_allowable_acceleration)
+    Polynomial2dOrder3_t path, const unsigned int num_elements, const float t_start,
+    const float t_end, const float max_allowable_acceleration,
+    float* max_allowable_speed_profile)
```

Comment:
looks  like this is unused

Reply:
Good catch, that's part if the old version. Removed 👍